python/ray/data/_internal/execution/operators/map_data_processor.py

```diff
@@ -0,0 +1,317 @@
+import functools
+import itertools
+from enum import Enum
+from typing import Any, Callable, Dict, Iterable, List, Optional, Union
+
+from ray.data._internal.block_batching.block_batching import batch_blocks
+from ray.data._internal.execution.interfaces.task_context import TaskContext
+from ray.data._internal.output_buffer import BlockOutputBuffer
+from ray.data.block import Block, BlockAccessor, DataBatch
+from ray.data.context import DataContext
+
+Row = Dict[str, Any]
+MapTransformFnData = Union[Block, Row, DataBatch]
+MapTransformCallable = Callable[
+    [Iterable[MapTransformFnData], TaskContext], Iterable[MapTransformFnData]
+]
+BlockTransformCallable = Callable[[Iterable[Block], TaskContext], Iterable[Block]]
+RowTransformCallable = Callable[[Iterable[Row], TaskContext], Iterable[Row]]
+BatchTransformCallable = Callable[
+    [Iterable[DataBatch], TaskContext], Iterable[DataBatch]
+]
+InitFn = Callable[[], None]
+
+
+class MapTransformFnDataType(Enum):
+    """An enum that represents the input/output data type of a MapTransformFn."""
+
+    Block = 0
+    Row = 1
+    Batch = 2
+
+
+class MapTransformFn:
+    """Represents a single transform function in a MapDataProcessor."""
+
+    def __init__(
+        self,
+        callable: MapTransformCallable,
+        input_type: MapTransformFnDataType,
+        output_type: MapTransformFnDataType,
+    ):
+        """
+        Args:
+            callable: the underlying Python callable object.
+            input_type: the type of the input data.
+            output_type: the type of the output data.
+        """
+        self._callable = callable
+        self._input_type = input_type
+        self._output_type = output_type
```

Comment:
does input type need to be the same as output type?

Reply:
no. The whole data processor needs to take Blocks and output Blocks. But the intermediate types can be Block/Row/Batch.